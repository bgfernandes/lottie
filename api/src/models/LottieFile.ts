import { JSONSchema7 } from 'json-schema'
import BaseModel from './BaseModel'

export default class LottieFile extends BaseModel {
  id: string | undefined
  slug: string | undefined
  url: string | undefined

  static get tableName():string {
    return 'lottie_files'
  }

  static get jsonSchema():JSONSchema7 {
    return {
      type: 'object',

      properties: {
        id: { type: 'string' },
        slug: { type: 'string', minLength: 21, maxLength: 21 },
        url: { type: 'string' },
        created_at: { type: 'string' },
        updated_at: { type: 'string' }
      },

      required: ['slug', 'url']
    }
  }
}
