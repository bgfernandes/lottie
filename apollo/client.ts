
import { ApolloClient, InMemoryCache, NormalizedCacheObject } from '@apollo/client'

let apolloClient: ApolloClient<NormalizedCacheObject>

if (typeof window === 'undefined') {
  // code is running in the server side
  const { SchemaLink } = require('@apollo/client/link/schema')
  const { schema } = require('./schema')

  apolloClient = new ApolloClient({
    ssrMode: true,
    link: new SchemaLink({ schema }),
    cache: new InMemoryCache(),
  })
} else {
  // code is running in the browser
  const { HttpLink } = require('@apollo/client/link/http')

  apolloClient = new ApolloClient({
    ssrMode: false,
    link: new HttpLink({
      uri: '/api/graphql',
      credentials: 'same-origin',
    }),
    cache: new InMemoryCache(),
  })
}

export default apolloClient
