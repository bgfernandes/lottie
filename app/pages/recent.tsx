import { GetServerSideProps, NextPage } from 'next'
import LottieFilesList, {
  LOTTIE_FILES_LIST_PAGE_SIZE,
  LOTTIE_FILES_QUERY,
} from '../components/LottieFilesList'
import { initializeApollo } from '../lib/apollo-client'
import { LottieFile } from '../types'
import Container from './../components/Container'

export const getServerSideProps: GetServerSideProps = async (_context) => {
  const apolloClient = initializeApollo()

  const { data } = await apolloClient.query({
    query: LOTTIE_FILES_QUERY,
    variables: { offset: 0, limit: LOTTIE_FILES_LIST_PAGE_SIZE },
  })

  return {
    props: {
      firstPageLotties: data.lottieFiles.lotties,
      initialApolloState: apolloClient.cache.extract(),
    },
  }
}

type RecentPageProps = {
  firstPageLotties: [LottieFile]
}

const RecentPage: NextPage<RecentPageProps> = ({ firstPageLotties }) => {
  return (
    <Container>
      <LottieFilesList firstPageLotties={firstPageLotties} />
    </Container>
  )
}

export default RecentPage
