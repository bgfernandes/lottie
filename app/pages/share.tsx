import { NextPage } from 'next'
import { useRouter } from 'next/router'
import { UploadLottieFile } from '../components/UploadLottieFile'

const Share: NextPage = () => {
  const router = useRouter()

  const onUpload = (slug: string) => {
    router.push('lotties/' + slug)
  }

  return <UploadLottieFile onUpload={onUpload} />
}

export default Share
