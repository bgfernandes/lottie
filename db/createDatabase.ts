/*
  Helper script file to create the DB
*/

import knex from 'knex'
import dotenv from 'dotenv'
import path from 'path'

if (require.main === module) {
  dotenv.config({
    path: path.resolve(__dirname, '..', '.env.local')
  })

  createDatabase()
}

export default function createDatabase():Promise<void> {
  const knexInstance = knex({
    client: 'pg',
    connection: {
      host : process.env.DB_HOST,
      user : process.env.DB_USER,
      password : process.env.DB_PASS
    }
  })

  return knexInstance.raw(`CREATE DATABASE ${process.env.DB_DATABASE};`)
    .then(() => {
      console.log(`Created Database ${process.env.DB_DATABASE}`)
    })
    .catch((err) => {
      if (err.message.match(/^CREATE DATABASE .*; - database ".*" already exists$/)) {
        console.log(`Database ${process.env.DB_DATABASE} already exists.`)
      } else {
        console.error(`Error creating database ${process.env.DB_DATABASE}`)
        console.error(err)
      }
    })
    .finally(() => {
      knexInstance.destroy()
    })
}
