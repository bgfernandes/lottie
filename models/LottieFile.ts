import { JSONSchema7 } from 'json-schema'
import BaseModel from './BaseModel'

export default class LottieFile extends BaseModel {
  id: string | undefined;

  static get tableName():string {
    return 'lottie_files'
  }

  static get jsonSchema():JSONSchema7 {
    return {
      type: 'object',

      properties: {
        id: { type: 'string' },
        created_at: { type: 'string' },
        updated_at: { type: 'string' }
      }
    }
  }
}
