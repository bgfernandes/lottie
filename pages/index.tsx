import type { NextPage, GetServerSideProps } from 'next'
import { gql, useQuery } from '@apollo/client'
import { initializeApollo } from '../apollo/client'

const HELLO_MESSAGE_QUERY = gql`
  query HelloMessage {
    hello {
      message
    }
  }
`

export const getServerSideProps: GetServerSideProps = async (_context) => {
  const apolloClient = initializeApollo()

  const { data } = await apolloClient.query({
    query: HELLO_MESSAGE_QUERY,
  })

  return {
    props: {
      serverSideData: data,
      initialApolloState: apolloClient.cache.extract()
     },
  }
}

type HomeProps = {
  serverSideData: {
    hello: {
      message: String
    }
  }
}

const Home: NextPage<HomeProps> = ({ serverSideData }) => {
  const { data } = useQuery(HELLO_MESSAGE_QUERY)
  return (
    <div>
      Hello world.

      <p className="flex justify-center">
        This page is using tailwindcss.
      </p>

      <p>
        This message was fetched from the graphql API on the server side: { serverSideData.hello.message }
      </p>

      <p>
        This message was fetched from the graphql API on the client side: { data ? data.hello.message : 'loading' }
      </p>
    </div>
  )
}

export default Home
