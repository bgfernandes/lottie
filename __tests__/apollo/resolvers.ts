/*
  Integration tests for the Apollo Server, uses the DB
*/

import { ApolloServer } from 'apollo-server-micro'
import { loadEnvConfig } from '@next/env'
import { Knex } from 'knex'
import knexCleaner from 'knex-cleaner'
import { gql } from '@apollo/client'
import { Model } from 'objection'
import moment from 'moment'
import createApolloServer from '../../apollo/server'
import { HELLO_MESSAGE_QUERY } from '../../pages/index'
import createDatabase from '../../db/createDatabase'
import createKnexInstance from '../../db/knexInstance'
import LottieFile from '../../models/LottieFile'

const projectDir = process.cwd()
loadEnvConfig(projectDir)

let server: ApolloServer
let knex: Knex


describe('Resolvers', () => {
  beforeAll(async () => {
    // create test database if not already created, run migrations, and setup objection
    await createDatabase()
    knex = createKnexInstance()
    await knex.migrate.latest({
      directory: './db/migrations'
    })
    Model.knex(knex)

    server = createApolloServer()
  })

  afterAll(async () => {
    await knex.destroy()
  })

  beforeEach(async () => {
    await knexCleaner
      .clean(knex, { ignoreTables: ['knex_migrations', 'knex_migrations_lock'] })
  })

  it('runs the home query successfully', async () => {
    const res = await server.executeOperation({ query: HELLO_MESSAGE_QUERY })
    expect(res).toMatchSnapshot()
  })

  it('runs the lottieFiles query successfuly and returns data', async () => {
    await LottieFile.query().insert({
      id: '1',
      createdAt: moment('2021-09-01'),
      updatedAt: moment('2021-09-02')
    })
    await LottieFile.query().insert({
      id: '2',
      createdAt: moment('2021-09-03'),
      updatedAt: moment('2021-09-04')
    })

    const LOTTIE_FILES_QUERY = gql`
      query {
        lottieFiles {
          id,
          createdAt,
          updatedAt
        }
      }
    `
    const res = await server.executeOperation({ query:  LOTTIE_FILES_QUERY})
    expect(res).toMatchSnapshot()
  })
})
