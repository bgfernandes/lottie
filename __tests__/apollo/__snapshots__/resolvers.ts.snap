// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers runs the home query successfully 1`] = `
Object {
  "data": Object {
    "hello": Object {
      "message": "Hello from Apollo.",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolvers runs the lottieFiles query successfuly and returns data 1`] = `
Object {
  "data": Object {
    "lottieFiles": Array [
      Object {
        "createdAt": "1630425600000",
        "id": "1",
        "updatedAt": "1630512000000",
      },
      Object {
        "createdAt": "1630598400000",
        "id": "2",
        "updatedAt": "1630684800000",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
